File: src/main/java/com/canefe/story/NPCContextGenerator.java
package com.canefe.story;

import org.bukkit.Bukkit;

import java.util.*;

public class NPCContextGenerator {

    private static NPCContextGenerator instance;
    private final Story plugin;

    private NPCContextGenerator(Story plugin) {
        this.plugin = plugin;
    }

    public static NPCContextGenerator getInstance(Story plugin) {
        if (instance == null) {
            instance = new NPCContextGenerator(plugin);
        }

        return instance;
    }

    public String generateDefaultContext(String npcName, String role, int hours, int minutes, String season, String date) {
        Random random = new Random();

        // Randomly select personality traits
        // plugin.getTraitList() is a method that returns a list of traits
        String trait = plugin.getTraitList().get(random.nextInt(plugin.getTraitList().size()));
        String quirk = plugin.getQuirkList().get(random.nextInt(plugin.getQuirkList().size()));
        String motivation = plugin.getMotivationList().get(random.nextInt(plugin.getMotivationList().size()));
        String flaw = plugin.getFlawList().get(random.nextInt(plugin.getFlawList().size()));
        String tone = plugin.getToneList().get(random.nextInt(plugin.getToneList().size()));


        // Construct personality description
        String personality = " is " + trait + ", has the quirk of " + quirk +
                ", is motivated by " + motivation + ", and their flaw is " + flaw +
                ". They speak in a " + tone + " tone.";

        // Construct the context
        return  npcName + personality +
                " The time is " + hours + ":" + String.format("%02d", minutes) + " in the " + season;

    }

    public String updateContext(String context, String npcName, int hours, int minutes, String season, String date) {
        Bukkit.getLogger().info("Updating context for NPC: " + npcName);
        context = context.replaceAll("The time is \\d{1,2}:\\d{2}", "The time is " + hours + ":" + String.format("%02d", minutes));
        context = context.replaceAll("in the \\w+", "in the " + season);
        context = context.replaceAll("The date is \\d{4}-\\d{2}-\\d{2}", "The date is " + date);
        return context;
    }
}

File: src/main/kotlin/com/canefe/story/npc/NPCContextGenerator.kt
package com.canefe.story.npc

import com.canefe.story.Story
import com.canefe.story.api.StoryNPC
import com.canefe.story.api.character.AICharacter
import com.canefe.story.api.character.Character
import com.canefe.story.api.character.CharacterRecord
import com.canefe.story.api.character.PlayerCharacter
import com.canefe.story.npc.data.NPCContext
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import java.io.File
import kotlin.random.Random

/**
 * Service responsible for generating and updating NPC context information.
 */
class NPCContextGenerator(
    private val plugin: Story,
) {
    private fun generateDefaultContext(name: String): String {
        val random = Random(System.currentTimeMillis())
        val trait = plugin.config.traitList.random(random)
        val quirk = plugin.config.quirkList.random(random)
        val motivation = plugin.config.motivationList.random(random)
        val flaw = plugin.config.flawList.random(random)
        val tone = plugin.config.toneList.random(random)
        return "$name is $trait, has the quirk of $quirk, " +
            "is motivated by $motivation, and their flaw is $flaw. " +
            "They speak in a $tone tone."
    }

    // ── Public API ──────────────────────────────────────────────────────

    /**
     * Primary entry point: get or create context for a Character.
     */
    fun getOrCreateContextForNPC(character: Character): NPCContext? {
        val npc = if (character is AICharacter) character.npc else null
        val player = if (character is PlayerCharacter) character.player else null
        return getOrCreateContextInternal(
            displayName = character.name,
            entityId = character.entityId.toString(),
            npc = npc,
            player = player,
        )
    }

    /**
     * Get or create context for a StoryNPC.
     */
    fun getOrCreateContextForNPC(npc: StoryNPC): NPCContext? =
        getOrCreateContextInternal(
            displayName = npc.name,
            entityId = npc.uniqueId.toString(),
            npc = npc,
        )

    // ── Internal ────────────────────────────────────────────────────────

    private fun getOrCreateContextInternal(
        displayName: String,
        entityId: String?,
        npc: StoryNPC? = null,
        player: Player? = null,
    ): NPCContext? {
        try {
            val record: CharacterRecord? =
                when {
                    npc != null -> plugin.characterRegistry.getByStoryNPC(npc)
                    else -> plugin.characterRegistry.getByName(displayName)
                }

            // Resolve location
            var location = plugin.locationManager.getOrCreateDefaultLocation()
            if (npc != null && npc.isSpawned && npc.entity != null) {
                val npcLocation = plugin.locationManager.getLocationByPosition2D(npc.entity!!.location)
                if (npcLocation != null) location = npcLocation
            }

            val avatar = record?.let { plugin.characterRegistry.getMinecraftConfig(it.id)?.avatar ?: "" } ?: ""

            // Get relationships
            val relationships = plugin.relationshipManager.getAllRelationships(displayName)

            return NPCContext(
                name = record?.name ?: displayName,
                role = "",
                context = generateDefaultContext(record?.name ?: displayName),
                appearance = record?.appearance ?: "",
                location = location,
                avatar = avatar,
                memories = emptyList(),
                relationships = relationships,
                customVoice = record?.customVoice,
            )
        } catch (e: Exception) {
            plugin.logger.warning("Error while updating NPC context: ${e.message}")
            e.printStackTrace()
            return null
        }
    }

    // ── General contexts ────────────────────────────────────────────────

    private val generalContexts: MutableList<String> = mutableListOf()
    private val generalContextFile = File(plugin.dataFolder, "general-contexts.yml")

    init {
        loadGeneralContexts()
    }

    fun loadConfig() {
        loadGeneralContexts()
    }

    private fun loadGeneralContexts() {
        generalContexts.clear()
        if (!generalContextFile.exists()) return
        val config = YamlConfiguration.loadConfiguration(generalContextFile)
        generalContexts.addAll(config.getStringList("contexts"))
    }

    private fun saveGeneralContexts() {
        val config = YamlConfiguration()
        config.set("contexts", generalContexts)
        config.save(generalContextFile)
    }

    fun getGeneralContexts(): List<String> = generalContexts.toList()

    fun addGeneralContext(context: String) {
        if (!generalContexts.contains(context)) {
            generalContexts.add(context)
            saveGeneralContexts()
        }
    }

    fun removeGeneralContext(context: String) {
        if (generalContexts.remove(context)) {
            saveGeneralContexts()
        }
    }
}
